src/Network/Ricochet.hs

```diff
@@ -1,10 +1,58 @@
 module Network.Ricochet where
 
-import Network.Ricochet.Monad
+import Network.Ricochet.Monad hiding (socksPort)
 import Network.Ricochet.Types
 import Network.Ricochet.Connection
 import Network.Ricochet.Version
 import Network.Ricochet.Util
 
-import Network (PortID(..))
-import Control.Monad.State (runStateT)
+import Data.Map (fromList)
+import Data.Base32String (Base32String)
+import Data.ByteString (ByteString)
+import Data.Word (Word8)
+import Network (PortID(..), Socket, listenOn)
+import Control.Monad (void, forever)
+import Control.Monad.IO.Class (liftIO)
+import Control.Monad.State (evalStateT)
+import Network.Anonymous.Tor (mapOnion, withSession)
+
+-- | Start an action inside the Ricochet monad.
+--
+--   This function uses a Tor control socket to create the hidden service.
+--   If no key is supplied, the hidden service will be random
+startRicochet :: PortID           -- ^ Port to listen on
+              -> Maybe ByteString -- ^ RSA1024 private key in base64 encoding
+              -> PortID           -- ^ The Tor control port
+              -> [Contact]        -- ^ A list of known 'Contact's
+              -> PortID           -- ^ The port of the Tor SOCKS5 proxy
+              -> [(Word8, Connection -> Ricochet ())] -- ^ A list of version identifiers and their
+                                                      --   corresponding 'Connection' handlers
+              -> (Base32String -> Ricochet ()) -- ^ The action will be supplied the address of the
+                                               --   hidden service
```

Comment:
I don’t like this part, as it would be a lot more useful to have the address insite the Connection handlers.  Maybe we should add it to the `RicochetState`?  A `ReaderT` on the stack might be better.